### PR TITLE
Add kubeflow cluster environment

### PR DESCRIPTION
**Pull request by neggert**
## What does this PR do?

Adds a `ClusterEnvironment` that works with the [PyTorchJob operator in Kubeflow](https://www.kubeflow.org/docs/components/training/pytorch/).

One open question from our discussion in Slack: is there a way to automatically tell if we're running inside a PyTorchJob? I've looked into this a bit, and I don't think there is. It's pretty easy to tell if we're in Kubernetes (various environment variables), but I haven't found a good way to check whether we're specifically running in a PyTorchJob. We could **maybe** query the Kubernetes API to find out, but that seems like overkill to me. Looking for feedback here.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
    * Discussed in Slack: https://pytorch-lightning.slack.com/archives/CU5PF2E2J/p1619447878081900
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
Hey! Awesome

Just fyi here is where we select the environment:

https://github.com/PyTorchLightning/pytorch-lightning/blob/490cc57809ebeba19003b4101393a8a058217c31/pytorch_lightning/trainer/connectors/accelerator_connector.py#L482

And here the example of TorchElastic spying on Env variables to identify itself:
As you can see in TorchElasticEnvironment.is_using_torch_elastic() we spy on ENV variables:
https://github.com/PyTorchLightning/pytorch-lightning/blob/490cc57809ebeba19003b4101393a8a058217c31/pytorch_lightning/plugins/environments/torchelastic_environment.py#L29

Now, as you say we need to detect PyTorchJob. I don't know PyTorchJob, but I searched GitHub for os.environ query and found here https://github.com/kubeflow/pytorch-operator/search?q=os.environ there is a PYTORCHJOB_VERSION env variable. But I'm not sure if this one is guaranteed to be set.

(btw the coverage is sometimes red because not yet all jobs have reported results (e.g. GPU still running)

**Comment by neggert** [comment]
I'm not seeing `PYTORCHJOB_VERSION` get set in our Kubeflow install. It looks to me like that variable is used by the Python SDK, but it's not actually injected into the container itself.

Here's the code that injects variables into the pod containers: https://github.com/kubeflow/pytorch-operator/blob/4aeb6503162465766476519339d3285f75ffe03e/pkg/controller.v1/pytorch/pod.go#L259

The only thing I can think would be to look for `KUBERNETES_PORT`, `MASTER_ADDR`, `MASTER_PORT`, `WORLD_SIZE` and `RANK` to be set, but *not* `GROUP_RANK`, `LOCAL_RANK`, or `LOCAL_WORLD_SIZE`, since the latter would be set if running under TorchElastic. This seems like it would be brittle, but maybe I'm overthinking it.

**Comment by awaelchli** [comment]
You think it may be brittle because you suspect there are other cluster managers that could be running with Kubernetes and be sharing the same env variables? I wouldn't know how to answer this but I could ask around. If we are unsure, it may be better to go with the manual selection / through argument parsing.

**Comment by neggert** [comment]
Yeah, that's pretty much the lines I'm thinking along. If you're comfortable with it, though, I'm happy to make the code change.

**Comment by awaelchli** [comment]
yes sounds good !

File: pytorch_lightning/plugins/environments/kubeflow_environment.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    def set_global_rank(self, rank: int) -> None:
+        log.debug(
+            "KubeflowEnvironment.set_global_rank was called, but setting global rank is not allowed. Ignored."
+        )
+
+    def local_rank(self) -> int:
+        return 0
+
+    def node_rank(self) -> int:
+        return self.global_rank()
```
Comment (Borda) [inline pytorch_lightning/plugins/environments/kubeflow_environment.py:57]: @awaelchli just thinking why are they not as property?

Comment (awaelchli) [reply pytorch_lightning/plugins/environments/kubeflow_environment.py:57]: not my decision. they were never properties from the very beginning and when I started working on these environments it was easier to keep the existing pattern.

I will refactor it, issue tracking here #6303

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 2]
```diff
@@ -417,7 +420,14 @@ def select_training_type_plugin(self) -> TrainingTypePlugin:
...
                 ddp_plugin_cls = DDPSpawnShardedPlugin
-            elif use_ddp_cpu_slurm or use_slurm_ddp or use_ddp_cpu_torch_elastic or use_torchelastic_ddp:
+            elif (
+                    use_ddp_cpu_slurm or
+                    use_slurm_ddp or
+                    use_ddp_cpu_torch_elastic or
+                    use_torchelastic_ddp or
+                    use_kubeflow_ddp or
+                    use_ddp_cpu_kubeflow
+            ):
```
Comment (neggert) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:426]: This seems like it might be due for a refactor, but I just extended the existing code for now, since I'm not 100% sure of all the edge cases this might be handling.

File: tests/accelerators/test_accelerator_connector.py [hunk 3]
```diff
@@ -277,6 +282,80 @@ def on_fit_start(self, trainer, pl_module):
...
+
+    class CB(Callback):
+
+        def on_fit_start(self, trainer, pl_module):
+            assert trainer.use_ddp
+            assert isinstance(trainer.accelerator, GPUAccelerator)
+            assert isinstance(trainer.training_type_plugin, DDPPlugin)
+            assert isinstance(trainer.training_type_plugin.cluster_environment, KubeflowEnvironment)
+            assert trainer.training_type_plugin.cluster_environment.local_rank() == 0
+            assert trainer.training_type_plugin.task_idx == 1
```
Comment (awaelchli) [inline tests/accelerators/test_accelerator_connector.py:308]: task_idx is an old unused variable, in your case it will be local_rank so here you have to assert = 0
```suggestion
            assert trainer.training_type_plugin.task_idx == 0
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:426]: it's getting a bit ridiculous yes. 
I have ideas how to simplify this in the future. But your addition looks solid!
So to summarize: The Kubeflow cluster environment is available with: DDPPlugin, DDPSpawnPlugin, user custom plugin, correct?

**Review by awaelchli — APPROVED** [review]
great addition!!

**Review by carmocca — APPROVED** [review]

**Comment by carmocca** [comment]
> Any idea why that pre-commit check is failing? Pre-commit seems to run find on my end.

Unrelated, opened #7500 to fix it
